strawberry/types/types.py

```diff
@@ -167,3 +184,30 @@ def is_implemented_by(self, root: Union[type, dict]) -> bool:
 
         # All field mappings succeeded. This is a match
         return True
+
+
+class WithStrawberryDefinition(Protocol):
+    __strawberry_definition__: StrawberryObjectDefinition
+
+
+def has_strawberry_definition(klass: Any) -> TypeGuard[Type[WithStrawberryDefinition]]:
+    if hasattr(klass, "__strawberry_definition__"):
+        return True
+    # Generics remove dunder members here
+    # https://github.com/python/cpython/blob/3a314f7c3df0dd7c37da7d12b827f169ee60e1ea/Lib/typing.py#L1152
+    if is_concrete_generic(klass):
+        concrete = klass.__origin__
+        if hasattr(concrete, "__strawberry_definition__"):
+            klass.__strawberry_definition__ = concrete.__strawberry_definition__
+            return True
+    return False
+
+
+def get_strawberry_definition(klass: Any) -> Optional[StrawberryObjectDefinition]:
+    if has_strawberry_definition(klass):
+        return klass.__strawberry_definition__
+    return None
+
+
+# TODO: remove when deprecating _type_definition
+TypeDefinition = StrawberryObjectDefinition
```

Comment:
Good to see the alias here, was worried about it for a second 😁

Reply:
I don't know if this would have other implications, but it would be nice to mark this as [deprecated](https://peps.python.org/pep-0702/) like this:

```python
if TYPE_CHECKING:
    @deprecated("Use StrawberryObjectDefinition instead")
    class TypeDefinition(StrawberryObjectDefinition):
        ...
else:
    TypeDefinition = StrawberryObjectDefinition
```

That decorator can be used with the latest version of `typing_extensions`: https://typing-extensions.readthedocs.io/en/latest/#deprecated